File: regex.js
// regex for * and .
function isMatch(s, p){

	if(p ===''){
		return true;
	}

	// contains *
	// search the through whole string for potential matches via recursion

	// if there are none, thats ok! because * means any or none. so do a recursive call to look for the rest of the pattern in the string
	if (p[1] == '*'){
		for (var i = 0; i < s.length && (p[0]==='.' || p[0]===s[0]); i++){
			if (isMatch(s.substr(i+1), p.substr(2))){
				return true;
			}
		}
		return isMatch(s, p.substr(2));

	}

	// if string is not empty and if character match... then we need to match up the rest of the string! so make a recursive call

	if (s!=='' && (p[0]==='.' || p[0]===s[0])){
		return isMatch(s.substr(1), p.substr(1));
	}else{
		return false;
	}
}

function regex(s,p){
	for (var j = 0; j<s.length; j++){
		if (isMatch(s.substr(j),p)){
			return true;
		}
	}
	return false;
}


var s = "aabc";
var p = "a*.c";

console.log(regex(s,p));

var s = "aabc";
var p = "a*.d";

console.log(regex(s,p));



var s = "aabc";
var p = "a*.d*";

console.log(regex(s,p));

File: regex.rb
#regex with . and *
def is_match?(str, p)
	# p "str: #{str} pattern: #{p} "
	return true if p.empty?

	if (p[1] == '*')
		i = 0
		while (i<str.length && (p[0]=='.' || p[0]== str[0]))
			return true if is_match?(str[i+1..-1], p[2..-1])
			i +=1
		end
		return is_match?(str, p[2..-1])
	end

	return (!str.empty? && (p[0] =='.' || p[0]==str[0])) ? is_match?(str[1..-1], p[1..-1]) : false
		# p "not empty and the first letter matches
end


def match(str, p)
	i = 0
	while i< str.length
		return true if is_match?(str[i..-1], p)
		i+=1
	end
	return false
end

p "match('abbc', '.b*c')"
p match("abbc", ".b*c")

p "match('abbc', 'ab*c')"
p match("abbc", "ab*c")

p "match('abbc', 'ab*d')"
p match("abbc", "ab*d")


p "match('aabc', 'a*.d*')"
p match("aabc", "a*.d*")
